sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala

```diff
@@ -60,6 +62,66 @@ case class BroadcastHashJoinExec(
     }
   }
 
+  override def outputPartitioning: Partitioning = {
+    joinType match {
+      case _: InnerLike =>
+        streamedPlan.outputPartitioning match {
+          case h: HashPartitioning => PartitioningCollection(expandOutputPartitioning(h))
+          case c: PartitioningCollection =>
+            def expand(partitioning: PartitioningCollection): Partitioning = {
+              PartitioningCollection(partitioning.partitionings.flatMap {
+                case h: HashPartitioning => expandOutputPartitioning(h)
+                case c: PartitioningCollection => Seq(expand(c))
+                case other => Seq(other)
+              })
+            }
+            expand(c)
+          case other => other
+        }
+      case _ => streamedPlan.outputPartitioning
+    }
+  }
+
+  // An one-to-many mapping from a streamed key to build keys.
+  private lazy val streamedKeyToBuildKeyMapping = {
+    val mapping = mutable.Map.empty[Expression, Seq[Expression]]
+    streamedKeys.zip(buildKeys).foreach {
+      case (streamedKey, buildKey) =>
+        val key = streamedKey.canonicalized
+        mapping.get(key) match {
+          case Some(v) => mapping.put(key, v :+ buildKey)
+          case None => mapping.put(key, Seq(buildKey))
+        }
+    }
+    mapping.toMap
+  }
+
+  // Expands the given partitioning by substituting streamed keys with build keys.
+  // For example, if the expressions for the given partitioning are Seq("a", "b", "c")
+  // where the streamed keys are Seq("b", "c") and the build keys are Seq("x", "y"),
+  // the expanded partitioning will have the following expressions:
+  // Seq("a", "b", "c"), Seq("a", "b", "y"), Seq("a", "x", "c"), Seq("a", "x", "y").
+  private def expandOutputPartitioning(partitioning: HashPartitioning): Seq[HashPartitioning] = {
+    def generateExprCombinations(
+        current: Seq[Expression],
+        accumulated: Seq[Expression],
+        all: mutable.ListBuffer[Seq[Expression]]): Unit = {
```

Comment:
Could you avoid using mutable in method arguments, and use a return value instead? e.g., https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala#L124-L133

Reply:
Changed. Thanks for the suggestion.